source/common/memory/stats.cc

```diff
@@ -2,121 +2,155 @@
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPhysicalBytes() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetProperties()["generic.physical_memory_used"].value;
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("generic.total_physical_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 void Stats::dumpStatsToLog() {
+#if defined(TCMALLOC)
+  ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", tcmalloc::MallocExtension::GetStats());
+#elif defined(GPERFTOOLS_TCMALLOC)
   constexpr int buffer_size = 100000;
   auto buffer = std::make_unique<char[]>(buffer_size);
   MallocExtension::instance()->GetStats(buffer.get(), buffer_size);
   ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", buffer.get());
+#else
+  return;
+#endif
 }
 
-} // namespace Memory
-} // namespace Envoy
-
+void AllocatorManager::tcmallocRelease() {
+#if defined(TCMALLOC)
+  tcmalloc::MallocExtension::ReleaseMemoryToSystem(bytes_to_release_);
+#elif defined(GPERFTOOLS_TCMALLOC)
+  MallocExtension::instance()->ReleaseToSystem(bytes_to_release_);
 #else
+  return;
+#endif
+}
 
-namespace Envoy {
-namespace Memory {
-
-uint64_t Stats::totalCurrentlyAllocated() { return 0; }
-uint64_t Stats::totalThreadCacheBytes() { return 0; }
-uint64_t Stats::totalCurrentlyReserved() { return 0; }
-uint64_t Stats::totalPageHeapUnmapped() { return 0; }
-uint64_t Stats::totalPageHeapFree() { return 0; }
-uint64_t Stats::totalPhysicalBytes() { return 0; }
-void Stats::dumpStatsToLog() {}
+/**
+ * Configures tcmalloc release rate from the page heap. If `bytes_to_release_`
+ * has been initialized to `0`, no heap memory will be released in background.
+ */
+void AllocatorManager::configureBackgroundMemoryRelease(Api::Api& api) {
+  RELEASE_ASSERT(!tcmalloc_thread_, "Invalid state, tcmalloc has already been initialised");
+  if (bytes_to_release_ > 0) {
+    tcmalloc_routine_dispatcher_ = api.allocateDispatcher(std::string(TCMALLOC_ROUTINE_THREAD_ID));
+    memory_release_timer_ = tcmalloc_routine_dispatcher_->createTimer([this]() -> void {
+      const uint64_t unmapped_bytes_before_release = Stats::totalPageHeapUnmapped();
+      tcmallocRelease();
+      const uint64_t unmapped_bytes_after_release = Stats::totalPageHeapUnmapped();
+      if (unmapped_bytes_after_release > unmapped_bytes_before_release) {
+        // Only increment stats if memory was actually released. As tcmalloc releases memory on a
+        // span granularity, during some release rounds there may be no memory released, if during
+        // past round too much memory was released.
+        // https://github.com/google/tcmalloc/blob/master/tcmalloc/tcmalloc.cc#L298
+        allocator_manager_stats_.released_by_timer_.inc();
+      }
+      memory_release_timer_->enableTimer(memory_release_interval_msec_);
+    });
+    tcmalloc_thread_ = api.threadFactory().createThread(
```

Comment:
Would it be possible to make this part of the main thread? I.e. add a timer to the main thread? It could be worthwhile to save a thread if all is needed is to run one timer.

Reply:
+1, while the thread only runs one timer, what really matters is the thread routine that timer invokes. I would not risk running tcmalloc memory release on main thread.

Reply:
@yanavlasov wdyt?